### DISCUSSION TITLE
FreeMarkerUtil. Error after upgrade to Keycloak 20

### DISCUSSION THREAD

**Discussion by agudzunas** [keycloak/keycloak#15438] [Archive]
Can't start keycloak after upgrade to version 20:

`ERROR: Unexpected error when starting the server in (production) mode
ERROR: Failed to start quarkus
ERROR: org/keycloak/theme/FreeMarkerUtil
ERROR: org.keycloak.theme.FreeMarkerUtil`

any ideas how to solve this?

**Comment by lexcao** [comment]
Any detailed logs?

**Comment by Carmezim** [comment]
bump

**Comment by agudzunas** [comment]
Sorry, I can't provide any logs right now. We will stick with v19 for a while. I et you know if I'll encounter this error in the future upgrades. Please close/delete this ticket.

**Comment by cedricguindon** [comment]
I'm having the same problem, but when upgrading my custom provider that uses org.keycloak.theme.FreeMarkerUtil

Looking at the docs, it looks like the Class was removed from 19 to 20. Any workaround?
That's not mentionned in the upgrade v19 to v20 release notes.

**Reply by alvinbaena** [reply]
I took a look around the code, and `org.keycloak.theme.FreeMarkerUtil` seems to have been replaced by `org.keycloak.theme.freemarker.FreeMarkerProvider`. It can be instantiated like any other Provider by using `KeycloakSession.getProvider`
